File: app/schemas/symbol_relationships.py
"""标的关联状态响应 Schema（WP1.5）。

为机会中心信息架构壳层提供富读模型：一次查询返回候选/观察/组合成员/持仓/告警
五类状态，前端 selector useSymbolRelationships(symbolId) 直接消费此结构。
"""
from __future__ import annotations

from pydantic import BaseModel, Field


class CandidateRelationship(BaseModel):
    """候选状态。"""
    has_candidate: bool = Field(False, description="是否为候选")
    candidate_id: int | None = None
    scope: str | None = None
    stage: str | None = None  # new/reviewed/watched/portfolio/excluded/expired
    action: str | None = None  # executable/caution/observe/reject
    priority_score: float | None = None
    quality_score: float | None = None
    timing_score: float | None = None
    data_credibility: str | None = None
    generated_at: str | None = None  # ISO 8601 datetime
    scan_run_id: int | None = None
    snapshot_id: int | None = None
    snapshot_generated_at: str | None = None


class ObservationRelationship(BaseModel):
    """观察项状态。"""
    has_observation: bool = Field(False, description="是否在观察池")
    watchlist_id: int | None = None
    watchlist_name: str | None = None
    watchlist_item_id: int | None = None
    origin_type: str | None = None  # manual/candidate/scan_result/alert/legacy_manual_unknown
    status: str | None = None  # watching/ready/invalid/archived
    priority: int | None = None
    tags: list[str] = Field(default_factory=list)
    target_portfolio_id: int | None = None
    added_at: str | None = None  # ISO 8601 datetime


class PortfolioMemberRelationship(BaseModel):
    """组合成员状态（第一阶段只读，预留字段）。"""
    has_portfolio_membership: bool = Field(False, description="是否为组合成员")
    portfolio_id: int | None = None
    portfolio_name: str | None = None
    member_id: int | None = None  # WP4 后填充
    member_status: str | None = None  # active/paused/archived
    execution_mode: str | None = None  # manual/confirm/auto
    source_type: str | None = None
    effective_from: str | None = None
    note: str | None = None


class PositionRelationship(BaseModel):
    """持仓状态。"""
    has_position: bool = Field(False, description="是否持仓")
    portfolio_id: int | None = None
    portfolio_name: str | None = None
    position_id: int | None = None
    quantity: float | None = None
    cost_price: float | None = None
    latest_price: float | None = None
    market_value: float | None = None
    opened_at: str | None = None  # ISO 8601 datetime


class AlertRelationship(BaseModel):
    """告警状态。"""
    has_active_alert: bool = Field(False, description="是否有激活的告警规则")
    alert_rule_ids: list[int] = Field(default_factory=list)
    active_alert_events: int = Field(0, description="未处理告警事件数")
    latest_alert_severity: str | None = None  # info/warning/error/critical
    latest_alert_at: str | None = None  # ISO 8601 datetime


class SymbolRelationships(BaseModel):
    """标的统一关联状态（WP1.5）。

    前端 selector useSymbolRelationships(symbolId) 使用此结构。
    今日决策、候选列表、组合页使用统一徽标。
    """
    symbol_id: int
    symbol: str | None = None
    candidate: CandidateRelationship = Field(default_factory=CandidateRelationship)
    observation: ObservationRelationship = Field(default_factory=ObservationRelationship)
    portfolio_member: PortfolioMemberRelationship = Field(default_factory=PortfolioMemberRelationship)
    position: PositionRelationship = Field(default_factory=PositionRelationship)
    alert: AlertRelationship = Field(default_factory=AlertRelationship)
    fetched_at: str  # ISO 8601 datetime
    degraded: bool = Field(False, description="接口部分失败时降级标记")
    degraded_reason: str | None = None


__all__ = [
    "CandidateRelationship",
    "ObservationRelationship",
    "PortfolioMemberRelationship",
    "PositionRelationship",
    "AlertRelationship",
    "SymbolRelationships",
]

File: frontend/src/types/symbolRelationships.ts
// WP1.5/WP1.2：标的统一关联状态 TS 类型定义
// 与后端 app/schemas/symbol_relationships.py 对齐
// 五段：candidate / observation / portfolio_member / position / alert

/** 候选状态 */
export interface CandidateRelationship {
  has_candidate: boolean;
  candidate_id: number | null;
  scope: string | null;
  // stage: new/reviewed/watched/portfolio/excluded/expired
  stage: string | null;
  // action: executable/caution/observe/reject
  action: string | null;
  priority_score: number | null;
  quality_score: number | null;
  timing_score: number | null;
  data_credibility: string | null;
  // ISO 8601 datetime
  generated_at: string | null;
  scan_run_id: number | null;
  snapshot_id: number | null;
  snapshot_generated_at: string | null;
}

/** 观察项状态 */
export interface ObservationRelationship {
  has_observation: boolean;
  watchlist_id: number | null;
  watchlist_name: string | null;
  watchlist_item_id: number | null;
  // origin_type: manual/candidate/scan_result/alert/legacy_manual_unknown
  origin_type: string | null;
  // status: watching/ready/invalid/archived
  status: string | null;
  priority: number | null;
  tags: string[];
  target_portfolio_id: number | null;
  // ISO 8601 datetime
  added_at: string | null;
}

/** 组合成员状态（第一阶段只读，预留字段） */
export interface PortfolioMemberRelationship {
  has_portfolio_membership: boolean;
  portfolio_id: number | null;
  portfolio_name: string | null;
  // WP4 后填充
  member_id: number | null;
  // active/paused/archived
  member_status: string | null;
  // manual/confirm/auto
  execution_mode: string | null;
  source_type: string | null;
  effective_from: string | null;
  note: string | null;
}

/** 持仓状态 */
export interface PositionRelationship {
  has_position: boolean;
  portfolio_id: number | null;
  portfolio_name: string | null;
  position_id: number | null;
  quantity: number | null;
  cost_price: number | null;
  latest_price: number | null;
  market_value: number | null;
  // ISO 8601 datetime
  opened_at: string | null;
}

/** 告警状态 */
export interface AlertRelationship {
  has_active_alert: boolean;
  alert_rule_ids: number[];
  active_alert_events: number;
  // info/warning/error/critical
  latest_alert_severity: string | null;
  // ISO 8601 datetime
  latest_alert_at: string | null;
}

/** 标的统一关联状态（前端 useSymbolRelationships hook 消费） */
export interface SymbolRelationships {
  symbol_id: number;
  symbol: string | null;
  candidate: CandidateRelationship;
  observation: ObservationRelationship;
  portfolio_member: PortfolioMemberRelationship;
  position: PositionRelationship;
  alert: AlertRelationship;
  // ISO 8601 datetime
  fetched_at: string;
  // 接口部分失败时降级标记
  degraded: boolean;
  degraded_reason: string | null;
}
